hare/eligibility/oracle.go

```diff
@@ -123,23 +126,50 @@ type vrfMessage struct {
 	Round  int32
 }
 
+func buildKey(l types.LayerID, r int32) [2]uint64 {
+	return [2]uint64{uint64(l), uint64(r)}
+}
+
 // buildVRFMessage builds the VRF message used as input for the BLS (msg=Beacon##Id##Layer##Round)
 func (o *Oracle) buildVRFMessage(layer types.LayerID, round int32) ([]byte, error) {
+	key := buildKey(layer, round)
+
+	// check cache
+	if val, exist := o.vrfMsgCache.Get(key); exist {
+		return val.([]byte), nil
+	}
+
+	o.lock.Lock()
```

Comment:
the cache itself isn't locked ?

Reply:
The cache is locked but we want to avoid double calculations.